Fix survey timings

clementzach opened the pull request:
Fix #81 

This implements survey delivery timings for absolute and relative timings. 

It also implements type checking and some PEP 8. 

I'm not sure what the process of reviewing code will be, but the commit history is probably pretty messy. If needed, I can go back and recreate these changes with commits that make more sense.

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +23,121 @@ def convert_time_to_date(submit_time, day, time):
+                current_time = datetime.datetime.fromisoformat(
+                    intervention_dict[t[0]]) + pd.Timedelta(
+                        t[1], unit='days') + pd.Timedelta(
+                            t[2], unit='seconds')  # seconds after time
```
```
                current_time = (
                    datetime.datetime.fromisoformat(intervention_dict[t[0]])
                    + pd.Timedelta(t[1], unit='days')
                    + pd.Timedelta(t[2], unit='seconds')
                )  # seconds after time
```

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,40 +146,84 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
-def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg, study_tz=None):
+                s_times = s_times + \
+                    generate_survey_times(
+                        time_start, time_end, timings=s['timings'],
```
`generate_survey_times()` expects strings as the first two arguments, so need to either update the function signature or cast to string

clementzach replied:
survey_submits was unnecessarily casting these as timestamps, so I deleted that and the conflict is removed.

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,40 +146,84 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+    else:
+        times_sur_df = pd.DataFrame(
+            columns=['delivery_time', 'next_delivery_time', 'id', 'beiwe_id'])
+    return(times_sur_df)
```
Remove redundant parens

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,40 +146,84 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+
+
+def survey_submits(
+        study_dir: str,
```
Param `study_dir` value is not used

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,40 +146,84 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+        beiwe_ids: list,
+        agg: pd.DataFrame,
+        study_tz: str = None,
+        all_interventions_dict: dict = {}
```
Default arg value is mutable
https://docs.python-guide.org/writing/gotchas/

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,40 +146,84 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+        time_end: str,
+        beiwe_ids: list,
+        agg: pd.DataFrame,
+        study_tz: str = None,
```
Param `study_tz` value is not used

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+            'next_delivery_time',
+            'survey id',
+            'beiwe_id',
+            'config_id'])['submit_flg'].max().reset_index()
```
`])['submit_flg'].max().reset_index()` should be on a separate line

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+    submit_lines3['submit_time'] = np.where(
+        submit_lines3.submit_flg == 1,
+        submit_lines3['Local time'],
+        np.array(
```
`np.array` call can fit on one line

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+
+    submit_lines_summary = pd.concat(
+        [num_surveys, num_complete_surveys, avg_time_to_submit], axis=1).\
+        reset_index()
```
```
    submit_lines_summary = pd.concat(
        [num_surveys, num_complete_surveys, avg_time_to_submit], axis=1
    ).reset_index()
```

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+    submit_lines3['submit_time'] = np.where(
+        submit_lines3['submit_flg'] == 1,
+        submit_lines3['submit_time'],
+        np.array(
```
`np.array` call can fit on one line

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+    submit_lines3['time_to_submit'] = np.where(
+        submit_lines3['submit_flg'] == 1,
+        submit_lines3['time_to_submit'],
+        np.array(
```
`np.array` call can fit on one line

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -143,77 +234,147 @@ def survey_submits(study_dir, config_path, time_start, time_end, beiwe_ids, agg,
+
+
+def survey_submits_no_config(study_dir: str,
+                             study_tz: Optional[str] = None) -> pd.DataFrame:
```
`study_dir` and `study_tz` are missing from docstring

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +384,43 @@ def summarize_submits(df):
+        index=[
+            'survey id',
+            'beiwe_id',
+            'surv_inst_flg'],
```
`index=[...]` can fit on one line

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +384,43 @@ def summarize_submits(df):
+    Extracts user intervention information for use in survey_timings.
+    Args:
+        filepath: the path to a table containing patient information
+        Note: right now, the user has to copy the table from the study website
```
`Note` is not a param of this function

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +384,43 @@ def summarize_submits(df):
+    filepath(str)
+    """
+    if filepath is None:
+        return({})  # empty dict
```
Remove redundant parens

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +384,43 @@ def summarize_submits(df):
+        for survey in full_dict[user].keys():
+            for time in full_dict[user][survey].keys():
+                output_dict[user][time] = full_dict[user][survey][time]
+    return(output_dict)
```
Remove redundant parens

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+def survey_stats_main(
+        study_folder: str,
+        output_folder: str,
+        tz_str: Optional[str] = "America/New_York",
```
Is this a useful default?

clementzach replied:
I believe around 70% of studies are done in our time zones, so I made this the default. But, perhaps users would expect UTC as default.

hackdna replied:
Since Forest will be released as a library for worldwide use I think it's better to assume UTC as a default.

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
 
+    """
+    if not os.path.isdir(output_folder):
+        os.mkdir(output_folder)
```
No need for a conditional, simply: `os.makedirs(output_folder, exist_ok=True)`
https://docs.python.org/3.8/library/os.html#os.makedirs

clementzach replied:
Ah, thanks!

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+            study_folder, tz_str, participant_ids)
+        if agg_data.shape[0] == 0:
+            print("Error: No survey data found")
+            return(True)
```
Redundant parens

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+                                            tz_str, participant_ids)
+        if agg_data.shape[0] == 0:
+            print("Error: No survey data found")
+            return(True)
```
Redundant parens

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+            os.path.join(
+                output_folder,
+                'answers_data.csv'),
+            index=False)
```
```
        ca_detail.to_csv(os.path.join(output_folder, 'answers_data.csv'),
                         index=False)
```

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+            os.path.join(
+                output_folder,
+                'answers_summary.csv'),
+            index=False)
```
```
        ca_summary.to_csv(os.path.join(output_folder, 'answers_summary.csv'),
                          index=False)
```

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+                time_end, participant_ids, agg_data,
+                tz_str, all_interventions_dict)
+            if ss_summary.shape[0] > 0:
+                ss_detail.to_csv(
```
This statement can fit on fewer lines

hackdna replied:
Still looks the same to me.

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+                        output_folder,
+                        'submits_data.csv'),
+                    index=False)
+                ss_summary.to_csv(
```
This statement can fit on fewer lines

clementzach replied:
Fixed. Thanks!

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+                print("An Error occurred when getting survey submit summaries")
     # Write out summaries
-    agg_data.to_csv(os.path.join(output_dir, 'agg_survey_data.csv'), index=False)
+    agg_data.to_csv(
```
This statement can fit on fewer lines

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
-    submits_tbl = survey_submits_no_config(study_dir, study_tz)
-    submits_tbl.to_csv(os.path.join(output_dir, 'submits_alt_summary.csv'), index=False)
+    submits_tbl = survey_submits_no_config(study_folder, tz_str)
+    submits_tbl.to_csv(
```
This statement can fit on fewer lines

clementzach replied:
Fixed. Thanks!

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,109 @@
+            output_folder,
+            'submits_alt_summary.csv'),
+        index=False)
+    return(True)
```
Redundant parens

hackdna reviewed (changes requested):
Thanks! Could you also add tests for new functions?

clementzach commented:
> Thanks! Could you also add tests for new functions?

Tests are added for the one function I added. I would probably have time to add tests for the rest of the functions next weekend.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -1,19 +1,20 @@
-def subset_answer_choices(answer):
+
+# Write a function that subsets the list if it starts with nan and ends
+# with two of the same elements
```
Looks like that comment was redundant with what the author already had in the docstring. I removed it.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -1,19 +1,20 @@
+def subset_answer_choices(answer: list) -> list:
     """
-    If a user changes their answers multiple times, an iOS device will have redundant answers at the beginning
+    Remove Duplicate Answers.
```
Ah, my linter was being overzealous. I updated those.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -29,13 +30,16 @@ def subset_answer_choices(answer):
-# Function that takes aggregated data and adds list of changed answers and first and last times and answers
-def agg_changed_answers(study_dir, config_path, agg, study_tz=None):
+# Function that takes aggregated data and adds list of changed answers and
+# first and last times and answers
```
Thanks for noticing that. I added that in.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -90,50 +110,85 @@ def agg_changed_answers_summary(study_dir, config_path, agg, study_tz=None):
+    most_common_answer = detail.groupby(summary_cols)['all_answers'].\
+        apply(lambda x: max(
+            set([x for x in x for x in x]),
+            key=[x for x in x for x in x].count, default=0))
```
I would need to familiarize myself more with their code more in order to refactor this. I'll leave this for now but hopefully I'll get back to this Monday morning when I am more able to do so.

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -1,22 +1,25 @@
 logger = logging.getLogger(__name__)
 
-
 # Modified from legacy beiwetools code
```
read_and_aggregate looks like it's sycamore-specific, so my guess is that it just applies to the read_json file. I wish I could ask whoever wrote the code :)

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -253,80 +293,107 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
-    tz = pytz.timezone(tz_str)
-    df_merged['Local time'] = [t.replace(tzinfo=None) for t in df_merged['Local time']]
+    df_merged['Local time'] = df_merged[utc_col].dt.tz_localize('UTC').dt.\
+        tz_convert('America/New_York').dt.tz_localize(None)
```
Thanks for catching that!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -253,80 +293,107 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
-def aggregate_surveys_config(study_dir, config_path, study_tz=None):
+def aggregate_surveys_config(study_dir: str,
+                             config_path: str,
+                             study_tz: Optional[str] = None,
```
Thanks for catching that! I meant to have it pass through to convert_timezone_df

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -383,22 +453,25 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
             try:
                 f = pd.read_csv(os.path.join(survey_dir, fp))
-            except:
+            except BaseException:
```
No idea what the author was meaning to do here. Adding this to the list of things to look at when I've had more sleep.

clementzach replied:
switched this to use continue instead of try/except, so this is fixed.

clementzach commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +23,121 @@ def convert_time_to_date(submit_time, day, time):
+
+        times_df = pd.DataFrame(timings)
+        times_df.columns = ['year', 'month', 'day', 'second']
+        surveys = pd.to_datetime(times_df).tolist()
```
to_datetime() can handle int, float, str, datetime, list, tuple, 1-d array, Series, DataFrame/dict-like
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html

times_df is a data frame as defined above

clementzach commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +23,121 @@ def convert_time_to_date(submit_time, day, time):
 
 
-def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+def gen_survey_schedule(
```
I believed I fixed this (in the function above)

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -383,22 +453,25 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
                 pass
 
             # Check whether participant uses iOS
             if 'event' in f.columns:  # iOS: last columnname == 'event'
```
Yep, I'll take a look at why they wrote that try/except and replace it with something better.

clementzach replied:
switched this to use continue instead of try/except, so this is fixed.

clementzach replied:
Thanks! I caught a few kinds of exceptions that it could throw.

clementzach commented:
@hackdna thanks so much for all the work you put in to this review! I really appreciate the attention to detail you put into making this code readable, and catching bugs that would have eventually come up if not fixed now.

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -94,109 +103,144 @@ def read_and_aggregate(study_dir, beiwe_id, data_stream):
+    # survey
+    all_data.loc[all_data['event'] ==
+                 'Survey first rendered and displayed to user',
+                 ['surv_inst_flg']] = 1
```
Less compact but more readable:
```
    all_data.loc[
        all_data['event'] == 'Survey first rendered and displayed to user',
        ['surv_inst_flg']
    ] = 1
```

clementzach replied:
Thanks! This is fixed

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -253,80 +284,94 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
     # Remove notification and expiration lines
-    df_merged = df_merged.loc[(~df_merged['question id'].isnull()) | (~df_merged['config_id'].isnull())]
+    df_merged = df_merged.loc[(~df_merged['question id'].isnull()) | (
+        ~df_merged['config_id'].isnull())]
```
More readable:
```
    df_merged = df_merged.loc[(~df_merged['question id'].isnull())
                              | (~df_merged['config_id'].isnull())]
```

clementzach replied:
Thanks!

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -1,18 +1,19 @@
+
```
No blank line necessary here.

clementzach replied:
Fixed that

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +23,117 @@ def convert_time_to_date(submit_time, day, time):
+        time_end: str,
+        timings: Optional[list] = None,
+        survey_type: str = 'weekly',
+        intervention_dict: Optional[dict] = None) -> list:
```
Closing and opening parentheses should match. Also, there's an extra indent for all args.

clementzach replied:
This is updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +23,117 @@ def convert_time_to_date(submit_time, day, time):
+        time_start: str,
+        time_end: str,
+        beiwe_ids: list,
+        all_interventions_dict: dict) -> pd.DataFrame:
```
Closing and opening parentheses should match.

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+                s_times = s_times + \
+                    generate_survey_times(
+                        time_start, time_end, timings=s['timings'],
+                        survey_type='weekly')
```
More readable:
```
                s_times = s_times + generate_survey_times(
                    time_start, time_end, timings=s['timings'],
                    survey_type='weekly'
                )
```

clementzach replied:
Thanks! This is updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +369,39 @@ def summarize_submits(df):
+    tmp = tmp.pivot(
+        index=['survey id', 'beiwe_id', 'surv_inst_flg'],
+        columns='level_3',
+        values='Local time').reset_index()
```
More compact and readable:
```
    tmp = tmp.pivot(index=['survey id', 'beiwe_id', 'surv_inst_flg'],
                    columns='level_3',
                    values='Local time').reset_index()
```

clementzach replied:
This is updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+        time_start: str,
+        time_end: str,
+        beiwe_ids: list,
+        agg: pd.DataFrame,
```
Missing param `agg` in docstring

clementzach replied:
This is updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+        time_start,
+        time_end,
+        beiwe_ids,
+        all_interventions_dict)
```
More compact:
```
    sched = gen_survey_schedule(config_path,
                                time_start,
                                time_end,
                                beiwe_ids,
                                all_interventions_dict)
```

clementzach replied:
This has been updated as well

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+        sched[['delivery_time', 'next_delivery_time',
+               'id', 'beiwe_id']].drop_duplicates(),
+        agg[['Local time', 'config_id', 'survey id', 'beiwe_id']
+            ].loc[agg.submit_line == 1].drop_duplicates(),
```
More readable:
```
        sched[
            ['delivery_time', 'next_delivery_time', 'id', 'beiwe_id']
        ].drop_duplicates(),
        agg[
            ['Local time', 'config_id', 'survey id', 'beiwe_id']
        ].loc[agg.submit_line == 1].drop_duplicates(),
```

clementzach replied:
Thanks! This is updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+            'survey id',
+            'beiwe_id',
+            'config_id'
+        ])['submit_flg'].max().reset_index()
```
`)['submit_flg'].max().reset_index()` should be on a new line to match the opening parenthesis

clementzach replied:
This has been updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,121 +142,224 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
     #     # Select appropriate columns
-    submit_lines3 = submit_lines3[['survey id', 'delivery_time', 'beiwe_id', 'submit_flg', 'submit_time']]
+    submit_lines3 = submit_lines3[[
+        'survey id', 'delivery_time', 'beiwe_id', 'submit_flg', 'submit_time']]
```
More readable:
```
    submit_lines3 = submit_lines3[
        ['survey id', 'delivery_time', 'beiwe_id', 'submit_flg', 'submit_time']
    ]
```

clementzach replied:
Thanks! updated

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -223,9 +369,39 @@ def summarize_submits(df):
+    Extracts user intervention information for use in survey_timings.
+    Args:
+        filepath: the path to a json file containing patient interventions
+        information (downloaded from the beiwe website)
```
Missing indent

clementzach replied:
Thanks!

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,92 @@
+    os.makedirs(output_folder, exist_ok=True)
+    if participant_ids is None:
+        participant_ids = [u for u in os.listdir(
+            study_folder) if not u.startswith('.') and u != 'registry']
```
More readable:
```
        participant_ids = [u
                           for u in os.listdir(study_folder)
                           if not u.startswith('.') and u != 'registry']
```

clementzach replied:
Thanks! Updated

hackdna commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,92 @@
+                    index=False)
+                ss_summary.to_csv(os.path.join(output_folder,
+                                               'submits_summary.csv'),
+                                  index=False)
```
More readable:
```
                ss_summary.to_csv(os.path.join(
                    output_folder, 'submits_summary.csv'
                ), index=False)
```

clementzach replied:
Updated this. I think it at least makes more sense for me to put the os.path.join on its own line so I can understand what the filepath will be

hackdna commented on forest/sycamore/tests/.ipynb_checkpoints/Untitled-checkpoint.ipynb:
```diff
@@ -0,0 +1,6 @@
+ "metadata": {},
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Does this file need to be under version control?

clementzach replied:
Nope, I didn't realize I had added it. Sorry about that

hackdna commented on forest/sycamore/tests/test_functions.py:
```diff
@@ -0,0 +1,50 @@
+from forest.sycamore.survey_config import gen_survey_schedule
+
+# Needed so that relative paths in the below file will work
+os.chdir(os.path.join("forest", "sycamore", "tests"))
```
I am not sure why this is necessary.

clementzach replied:
I need to look more into how relative paths work in pytest myself. But, when I did not include this line, pushing to GitHub would return an error.

hackdna replied:
But not when running the tests locally? What was the error?

clementzach replied:
You can see it [here](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true)

> forest/sycamore/tests/test_functions.py:7: in <module>
> [16](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:16)
>     assert get_all_interventions_dict("empty_intervention_data.json") == {}
> [17](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:17)
> forest/sycamore/survey_config.py:4[18](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:18): in get_all_interventions_dict
> 18
>     full_dict = read_json(filepath)
> [19](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:19)
> forest/sycamore/functions.py:28: in read_json
> [20](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:20)
>     with open(study_dir, 'r') as f:
> [21](https://github.com/onnela-lab/forest/runs/5348311431?check_suite_focus=true#step:9:21)
> E   FileNotFoundError: [Errno 2] No such file or directory: 'empty_intervention_data.json'

hackdna replied:
This is because you've used relative paths to JSON files. They will work only if you run `pytest` from the `tests` dir. The standard way to solve this is something like:
```
TEST_DATA_DIR = os.path.dirname(os.path.abspath(__file__))
```
then:
```
get_all_interventions_dict(os.path.join(TEST_DATA_DIR, "empty_intervention_data.json")
```

hackdna commented on forest/sycamore/tests/test_functions.py:
```diff
@@ -0,0 +1,50 @@
+
+    assert empty_dict == {}
+
+    interventions_dict = get_all_interventions_dict(
```
Each test should test one thing. I think this one can be split in two here, something like: `test_get_interventions_with_empty_input()` and `test_get_interventions_with_normal_input()`

clementzach replied:
Thanks! This is updated :)

hackdna commented on forest/sycamore/tests/test_functions.py:
```diff
@@ -0,0 +1,50 @@
+    assert pd.isnull(sample_agg_data.loc[0, "time_prev"])
+
+    assert all(pd.to_datetime(sample_agg_data['timestamp'], unit='ms') ==
+               sample_agg_data['UTC time'])
```
`all()` requires an iterable

clementzach replied:
Hmm. I replaced this with np.mean, which I think doesn't necessarily need one

hackdna reviewed (changes requested):
Thank you, this is good progress!

Sorry if I wasn't clear about this earlier but please let the reviewer resolve the conversations on the PR - that makes it easier to see if the changes were correct.

I think the automatic code formatter you have used is not doing a great job in many cases, particularly regarding breaking down long lines and matching the indentation of opening/closing brackets/parentheses, so it might be easier just to format the code by hand based on flake8 recommendations.

Finally, it'd be great to add more tests - let me know if you need any help with that.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -90,50 +110,85 @@ def agg_changed_answers_summary(study_dir, config_path, agg, study_tz=None):
+            detail.first_time == detail.last_time) & (
+            detail.q_internal_id != 1),
+        detail.last_time_1,
+        detail.first_time)
```
Thanks, updated that

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -253,80 +293,107 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
     if tz_str is None:
         tz_str = 'America/New_York'
```
This is done for all functions

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -410,18 +483,27 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
+                    subset=[
+                        'date_hour',
+                        'event'],
+                    keep='first')
```
Thanks. I'll go through and try to condense code.

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -76,64 +77,106 @@ def agg_changed_answers(study_dir, config_path, agg, study_tz=None):
+# Create a summary file that has survey, beiwe id, question id, average
+# number of changed answers, average time spent answering question
+
+def agg_changed_answers_summary(study_dir: str,
```
I believe I made the correct change. Can I clarify: when you say "Opening parenthesis should match the closing one", you mean that the closing parenthesis should be at the same level of indentation as the start of the function?

hackdna replied:
Yes, if you have a new line after the opening parenthesis then there should be a new line right before the closing one:
```
def foo(
    a
):
```
or no new lines:
```
def foo(a):
```

clementzach commented on forest/sycamore/changed_answers.py:
```diff
@@ -76,64 +77,106 @@ def agg_changed_answers(study_dir, config_path, agg, study_tz=None):
+        'num_answers',
+        'first_time',
+        'last_time',
+        'time_to_answer']
```
made this change as well

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -1,22 +1,24 @@
 import pandas as pd
-import pytz
+
+
```
Thanks! This is done

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -410,18 +457,27 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
                     try:
-                        submitted = f.loc[f['date_hour'] == timestamp, 'submitted'][0]
+                        submitted = f.loc[f['date_hour']
+                                          == timestamp, 'submitted'][0]
```
Thanks! I made that change

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -51,9 +54,8 @@ def make_lookup():
-def q_types_standardize(q, lkp=question_types_lookup):
-    """
-    Standardizes question types using a lookup function
+def q_types_standardize(q: str, lkp: dict = question_types_lookup) -> str:
```
This is fixed. 

In this case, the function isn't adding/removing anything from the dict that is passed, so I had let that stand. But, I suppose it is best practice to not do that.

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -94,109 +103,144 @@ def read_and_aggregate(study_dir, beiwe_id, data_stream):
         survey_data['beiwe_id'] = beiwe_id
-        survey_data['UTC time'] = survey_data['UTC time'].astype('datetime64[ns]')
+        survey_data['UTC time'] = survey_data['UTC time'].astype(
+            'datetime64[ns]')
```
This is fixed

clementzach commented:
@hackdna I believe everything you mentioned has been fixed.

Thanks for clarifying about resolving conversations.

Yes, the automatic code formatter was less than helpful. I've now switched to pycharm, and stopped auto-formatting on save so that should solve the problem. 

I added tests for every high-level function, so at least pytest will catch whether a function errors out. I still could do some more work checking the specific values generated by those functions, though. 

I'm going to generate another test set of data this week to test more edge cases.

hackdna reviewed (changes requested):
Thank you! GitHub was a bit glitchy today, so please comment again on any thread that I might have missed and please re-request the review when you are done with the next round of updates so I'd know when it's ready.

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -94,109 +104,147 @@ def read_and_aggregate(study_dir, beiwe_id, data_stream):
+    # Splitting up surveys where there appears to be a time gap or no submit
+    # line.
+    all_data.loc[all_data.time_diff > datetime.timedelta(hours=2), [
+        'surv_inst_flg']] = 1
```
Thanks!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -94,109 +104,147 @@ def read_and_aggregate(study_dir, beiwe_id, data_stream):
 
-    all_data['surv_inst_flg'] = all_data.groupby(['survey id', 'beiwe_id'])['surv_inst_flg'].cumsum()
+    all_data['surv_inst_flg'] = all_data.groupby(['survey id', 'beiwe_id'])[
+        'surv_inst_flg'].cumsum()
```
Thanks!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -253,80 +288,95 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
+        return agg_data
+    agg_data['submit_line'] = agg_data.apply(
+        lambda row: 1 if row['event'] in [
+            'User hit submit', 'submitted'] else 0, axis=1)
```
Thanks!

clementzach commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,89 @@
-        agg_data = aggregate_surveys_no_config(study_dir, study_tz)
-
+        agg_data = aggregate_surveys_no_config(
+            study_folder, tz_str)
```
Thanks!

clementzach commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,89 @@
+            if ss_summary.shape[0] > 0:
+                ss_detail.to_csv(
+                    os.path.join(output_folder, 'submits_data.csv'),
+                    index=False)
```
Thanks!

clementzach commented on forest/sycamore/sycamore_main.py:
```diff
@@ -1,48 +1,89 @@
+                    index=False)
+                ss_summary.to_csv(
+                    os.path.join(output_folder, 'submits_summary.csv'),
+                    index=False)
```
Thanks!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -443,12 +512,18 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
+                # Looks like if Androids have double events, you should take
+                # the   last
+                f = f.drop_duplicates(
+                    subset=['date_hour', 'question id'], keep='last')
```
Thanks!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -368,37 +420,46 @@ def get_survey_timings(person_ids: List[str], study_dir: str, survey_id: str):
+                continue
+            except pd.errors.ParserError:
+                print("Parse error")
+                continue
```
Yes, this makes sense. I've implemented this everywhere I had a print statement.

hackdna replied:
Thanks! However, logging statements is probably the only place where old-style string formatting should be used. See here for details: https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -209,35 +268,22 @@ def parse_surveys(config_path, answers_l=False):
-
+def convert_timezone_df(df_merged: pd.DataFrame,
+                        tz_str: str = "UTC",
+                        utc_col: str = 'UTC time') -> pd.DataFrame:
```
"Pick a rule and stick to it. When a string contains single or double quote characters, however, use the other one to avoid backslashes in the string. It improves readability."
https://www.python.org/dev/peps/pep-0008/#string-quotes

I know this hasn't been very consistent throughout the project so far but let's stick with double quotes as the default.

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -250,83 +296,103 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
-    tz = pytz.timezone(tz_str)
-    df_merged['Local time'] = [t.replace(tzinfo=None) for t in df_merged['Local time']]
+    df_merged['Local time'] = df_merged[utc_col].dt.tz_localize(
+        'UTC'
```
Double quotes

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +25,114 @@ def convert_time_to_date(submit_time, day, time):
+def gen_survey_schedule(
+        config_path: str, time_start: str, time_end: str, beiwe_ids: list,
+        all_interventions_dict: dict
+                        ) -> pd.DataFrame:
```
Sorry, I should have been more clear. It should be just:
```
def gen_survey_schedule(
    config_path: str, time_start: str, time_end: str, beiwe_ids: list,
    all_interventions_dict: dict
) -> pd.DataFrame:
```
or
```
def gen_survey_schedule(config_path: str,
                        time_start: str,
                        time_end: str,
                        beiwe_ids: list,
                        all_interventions_dict: dict) -> pd.DataFrame:
```
or
```
def gen_survey_schedule(
        config_path: str,
        time_start: str,
        time_end: str,
        beiwe_ids: list,
        all_interventions_dict: dict
) -> pd.DataFrame:
```

clementzach replied:
Thanks for the clarification!

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -250,83 +296,103 @@ def convert_timezone_df(df_merged, tz_str=None, utc_col='UTC time'):
-    Cleans aggregated data
+def aggregate_surveys_no_config(study_dir: str, study_tz: str = "UTC",
+                                users: list = None
+                                ) -> pd.DataFrame:
```
simply:
```
def aggregate_surveys_no_config(study_dir: str, study_tz: str = "UTC",
                                users: list = None) -> pd.DataFrame:
```

clementzach replied:
Thanks!

hackdna commented on forest/sycamore/survey_config.py:
```diff
@@ -99,133 +141,261 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
-                                    'avg_time_to_submit']
+    return (submit_lines3.sort_values(
+        ["survey id", "beiwe_id"]).drop_duplicates(),
+        submit_lines_summary)
```
more readable:
```
    return submit_lines3.sort_values(
        ["survey id", "beiwe_id"]
    ).drop_duplicates(), submit_lines_summary
```

clementzach replied:
Thanks!

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -51,193 +54,234 @@ def make_lookup():
-        all_files = glob.glob(os.path.join(st_path, '*/*.csv'))
-        # Sort file paths for when they're read in
+        all_files = glob.glob(os.path.join(st_path, "*/*.csv"))
+        # Sort file paths for when they"re read in
```
they're

hackdna commented on forest/sycamore/functions.py:
```diff
@@ -51,193 +54,234 @@ def make_lookup():
+                surv["config_id"] = i
+                surv["question_id"] = q["question_id"]
+                surv["question_text"] = q["question_text"]
+                surv["question_type"] = q["question_type"]
```
This dictionary creation could be rewritten as a dictionary literal

hackdna reviewed (changes requested):
Thanks! Getting there.

clementzach commented on forest/sycamore/survey_config.py:
```diff
@@ -22,75 +25,114 @@ def convert_time_to_date(submit_time, day, time):
+
+        for s in start_dates:
+            # Get the starting day of week
+            #         dow_s = s.weekday()
```
Nope, thanks for catching that

clementzach commented on forest/sycamore/survey_config.py:
```diff
@@ -99,133 +141,261 @@ def gen_survey_schedule(config_path, time_start, time_end, beiwe_ids):
+    filepath(str)
+    """
+    if filepath is None:
+        return {}  # empty dict
```
yes, that's right

clementzach commented on forest/sycamore/tests/test_functions.py:
```diff
@@ -0,0 +1,128 @@
+from forest.sycamore.survey_config import gen_survey_schedule, survey_submits
+from forest.sycamore.changed_answers import agg_changed_answers_summary
+
+LOCAL_DIR = os.path.dirname(os.path.abspath(__file__))
```
Thanks!

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -1,47 +1,50 @@
+events = read_json(os.path.join(this_dir, "events.json"))
+question_type_names = read_json(
+    os.path.join(this_dir, "question_type_names.json")
+)
```
Yes, I'm not sure why they did this that way. I suppose maybe this way allows for changing question types more easily if new ones are added, but it would be just as easy to update this dict as it would be to update the json file with new types.

clementzach commented on forest/sycamore/functions.py:
```diff
@@ -1,47 +1,50 @@
 this_dir = os.path.dirname(__file__)
-events = read_json(os.path.join(this_dir, 'events.json'))
-question_type_names = read_json(os.path.join(this_dir, 'question_type_names.json'))
+events = read_json(os.path.join(this_dir, "events.json"))
```
Yes, looks like that was just included in a blind copy/paste

hackdna reviewed (approved):
Thank you very much!
